fecgov/fec-cms
# [WIP]Shareable accurate data table links

## Summary (required)

- Resolves #4010
- Research issue: https://github.com/fecgov/fec-cms/issues/4316
- **Related user-submitted issues:**
https://github.com/fecgov/FEC/issues/10195
https://github.com/fecgov/FEC/issues/10194
https://github.com/fecgov/FEC/issues/8821

This PR  addresses two existing issues with some data tables that are causing inconsistencies between the URL querystring parameters  and the state of the filters panel (chosen filters)  - making is impossible to share/ bookmark accurate data table links.

1. **Issue:** Certain checkbox filters  (in a checkbox dropdown widget) are not added to the URL querystring parameters when  chosen, or do not remove from the querystring when manually removed in the filter panel or taglist buttons. 
     - **Fix**: Use the correct macro variable in `committee_types.jinja` for data-name attributes of fieldsets to ensure the checkbox add/remove works as expected for all checkboxes
2. **Issue:** Having checkboxes with differing `data-name` attributes inside a fieldset confuses existing logic that  compares all filters in the filter  panel against the filters  indicated by querystring parameters . This  causes some filters to be stripped from the URL and sometimes not be included in the filtered API call for data. Another symptom of this is that if there are two checkboxes with the same letter value, checking one can sometimes automatically check the other. 
    - **Fix:** Upon page load, programmatically activate all checkbox filters indicated in the querystring parameters (if they are not already activated  by existing logic)

This will ensure that the URL  matches the state of the filter panel and vice versa allowing datatable URLs to remain accurate when shared, bookmarked or refreshed.

More  to come...

### Required reviewers

(Include who is required to review prior to merge. For example: One designer and two front end developer reviews are required prior to merge)

## Impacted areas of the application

General components of the application that this PR will affect:

-  

## Screenshots

(Include a screenshot of the new/updated features in context (“in the wild”). If it is an interface change, include both before and after screenshots)

## Related PRs

Related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()

## How to test
This branch is deployed to feature space
- checkout branch and run npm run build-js
- Test  adding `Lobbyist/Registrant PAC` as a filter on http://127.0.0.1:8000/data/committees/pac-party/
   -  Does it get  added to the URL  as `committee_designation=B` ?
    - Does removing this via the filter panel or taglist buttons  also remove it from the URL?
- Test copy/pasting a URL with several query parameters into a page
    - Does it preserve the URL you pasted in ?
    - Are the filters in the  URL activated in the filter  panel/taglist?
    - Are the filter counts correct ?
 - Compare the data results returned when choosing filters manually with the results returned when pasting a link with same filters into browser
   - Is the data the same ?
   - Are the total data result counts the same ?
- Another test we  should do is to compare  the data returned from using [OpenFEC](https://api.open.fec.gov/) against  results from datatables to ensure the data is indeed filtered according to what  the URL and filter panel/tags indicate.
